### PR TITLE
Enhance CLI UX and Add Comprehensive Test Suite

### PR DESCRIPTION
**Pull request by azzazzel**
## Problem

This PR significantly improves the Pinecone CLI user experience and adds a comprehensive test suite. It introduces interactive mode for index creation, better defaults, improved command structure, and extensive testing infrastructure.

### CLI Enhancements for Index Creation Flows
- **Interactive Mode**: Added `--interactive` flag for guided index creation with sensible defaults
- **Better Defaults**: Index creation now defaults to serverless with reasonable configurations
- **Improved UX**: Changed `--name` flag to positional argument for more intuitive usage
- **Preview Mode**: Shows what will be created before confirmation
- **User Confirmation**: Added `-y/--yes` flag and confirmation prompts for safety

### Testing Infrastructure
- **BATS Framework**: Integrated BATS (Bash Automated Testing System) as Git submodules
- **Test Helpers**: Custom helper functions and JSON templates for validation
- **Comprehensive Tests**: Added 50+ test files covering CLI operations, edge cases, and constraints
- **OS-Agnostic**: Cross-platform test execution with proper dependency management
- **Installation Scripts**: OS-agnostic dependency installation (for test dependencies)

### Development Tools
- **`pcdev` Script**: Cross-platform development script that auto-detects OS/architecture
- **Auto-Build**: Automatically builds CLI when needed during development

### Documentation
- **Updated Contributing Guide**: Improved setup instructions and development workflow
- **Test Documentation**: Comprehensive test suite documentation with examples

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] Infrastructure change (CI configs, etc)
- [x] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

The PR includes the first tests for:
- Basic CLI operations and edge cases
- Index creation flows (serverless for now)
- Interactive mode workflows
- Constraint validation and error handling
- Deletion protection scenarios

See [test/README.md](test/README.md) for instructions how to run the tests

## Breaking Changes
- `-name` flag changed to positional argument in index create
- Some flag combinations now have stricter validation

### CODE REVIEW

File: pcdev [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+OS=$(uname -s | tr '[:upper:]' '[:lower:]')
+ARCH=$(uname -m)
+
+print_status "Detected OS: $OS, Architecture: $ARCH"
+print_status "Looking for binary in dist directory: $DIST_DIR"
+
+# Determine the correct binary path
+BINARY_PATH=""
+
+case "$OS" in
```
Comment (austin-denoble) [inline pcdev:146]: Mentioned above, but this approach feels quite brittle to me, on top of being entirely written in bash. What happens if goreleaser modifies the names of the paths in later versions? We then have to go through and refactor all of this.

File: pcdev [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    if [ ! -f "$CACHE_FILE" ]; then
+        return 0
+    fi
+    
+    # Check if any Go files have been modified since last build
+    local last_build_time=$(cat "$CACHE_FILE" 2>/dev/null || echo "0")
+    
+    # Use stat to get the most recent modification time of Go files
+    local latest_file_time=0
+    if command -v stat >/dev/null 2>&1; then
```
Comment (austin-denoble) [inline pcdev:91]: This seems a bit brittle depending on the OS the person is on. Building the binary with goreleaser is not a very expensive or time-consuming operation. I'm unclear on why we'd want to wrap this logic up like this rather than just having developers re-build as needed.

File: pcdev [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+print_error() {
+    # Always show errors, even in debug mode
+    echo -e "${RED}[pcdev]${NC} $1" >&2
+}
+
+# Get script directory
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+DIST_DIR="$SCRIPT_DIR/dist"
+CACHE_FILE="$SCRIPT_DIR/.pcdev_cache"
```
Comment (austin-denoble) [inline pcdev:72]: This file is only storing the last time a build was run, right?

File: pcdev [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        update_cache
+        print_status "Build completed successfully!"
+    fi
+fi
+
+# Check if dist directory exists (final check)
+if [ ! -d "$DIST_DIR" ]; then
+    print_error "dist directory not found. Please run 'goreleaser build --single-target --snapshot --clean' first."
+    exit 1
+fi
```
Comment (austin-denoble) [inline pcdev:134]: This feels odd to me because we're calling `goreleaser build --single-target --snapshot --clean` above, but then telling them to run it themselves here.

File: pcdev.ps1 [hunk 5]
```diff
@@ -0,0 +1,128 @@
+# pcdev.ps1 - PowerShell version for Windows compatibility
+# Development script to run the Pinecone CLI from dist folder
```
Comment (austin-denoble) [inline pcdev.ps1:2]: This is the same logic as above duplicated for windows powershell, right? Have you been able to test this yourself on a windows machine?

I could try and get my personal PC set up to test this, but I've not taken the time yet. Given we don't have any burning need for windows support in a development sense., I think we should just remove this entirely for now. It's a good thought though to account for various environments people might be working in.

I really do not have any familiarity with powershell, so I don't feel like I can review this effectively.

**Review by austin-denoble — COMMENTED** [review]
Focused mostly on the `pcdev` proposal in this first pass. Left questions and comments inline. Overall, I'm not enthusiastic about adding a large bash script file to manage some of these developer interactions, and I think there are more focused ways we could achieve some of these things. Happy to discuss things a bit more in our sync.

File: .github/workflows/build-and-test.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        run: |
+          cd test
+          chmod +x install-deps.sh
+          ./install-deps.sh
+
+      - name: Run tests
+        env:
+          SKIP_PCDEV_BUILD: "true"
+        run: |
+          ./pcdev config set-api-key $CLI_TEST_API_KEY
```
Comment (austin-denoble) [inline .github/workflows/build-and-test.yaml:47]: I saw the workflow is failing in CI:

```
  ./pcdev config set-api-key $CLI_TEST_API_KEY
  cd test
  ./bats -r -p -j 5 tests/
  shell: /usr/bin/bash -e {0}
  env:
    SKIP_PCDEV_BUILD: true
[pcdev] No suitable Linux binary found for architecture x86_64
-rwxr-xr-x 1 runner docker 23332028 Aug 20 16:34 /home/runner/work/cli/cli/dist/pc_linux_amd64_v1/pc
Error: Process completed with exit code 1.
```

This seems like an error `pcdev` is throwing. It seems like it should be using `pc_linux_amd64_v1`, which I see locally when I run goreleaser, but I'm not seeing it in any of the logic in the `pcdev` script when determining what to run linux on:

```
"linux")
        case "$ARCH" in
            "x86_64"|"amd64")
                # Try goreleaser v2 naming first (pc_Linux_x86_64)
                if [ -f "$DIST_DIR/pc_Linux_x86_64/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_Linux_x86_64/pc"
                # Fallback to older naming conventions
                elif [ -f "$DIST_DIR/pc_linux_amd64/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_linux_amd64/pc"
                elif [ -f "$DIST_DIR/pc_linux_x86_64/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_linux_x86_64/pc"
                fi
                ;;
            "aarch64"|"arm64"|"arm64_v8.0")
                # Try goreleaser v2 naming first (pc_Linux_arm64 or pc_Linux_arm64_v8.0)
                if [ -f "$DIST_DIR/pc_Linux_arm64_v8.0/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_Linux_arm64_v8.0/pc"
                elif [ -f "$DIST_DIR/pc_Linux_arm64/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_Linux_arm64/pc"
                # Fallback to older naming conventions
                elif [ -f "$DIST_DIR/pc_linux_arm64_v8.0/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_linux_arm64_v8.0/pc"
                elif [ -f "$DIST_DIR/pc_linux_arm64/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_linux_arm64/pc"
                fi
                ;;
            "armv7l"|"arm")
                # Try goreleaser v2 naming first (pc_Linux_arm or pc_Linux_arm_v7)
                if [ -f "$DIST_DIR/pc_Linux_arm_v7/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_Linux_arm_v7/pc"
                elif [ -f "$DIST_DIR/pc_Linux_arm/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_Linux_arm/pc"
                # Fallback to older naming conventions
                elif [ -f "$DIST_DIR/pc_linux_arm_v7/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_linux_arm_v7/pc"
                elif [ -f "$DIST_DIR/pc_linux_arm/pc" ]; then
                    BINARY_PATH="$DIST_DIR/pc_linux_arm/pc"
                fi
                ;;
        esac

        if [ -z "$BINARY_PATH" ]; then
            print_error "No suitable Linux binary found for architecture $ARCH"
            print_status "Available binaries:"
            ls -la "$DIST_DIR"/*/pc 2>/dev/null || echo "No binaries found"
            exit 1
        fi
        ;;
```

File: test/README.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
+# Test Suite Documentation
```
Comment (austin-denoble) [inline test/README.md:1]: I really do appreciate the level of detail and organization that went into getting this set up. I think the approaches to testing functionality individually can also be carried over to writing tests in Go. However, I don't think adding on a rather large and complex dependency that's written in and maintained in bash, is a good fit for a Go codebase.

One of my primary concerns here is we're now writing custom logic that needs to be maintained in bash, and a lot of these things like [delete_index_with_protection_handling](https://github.com/pinecone-io/cli/pull/32/files#diff-f6375fe66fbd9c9d7b89c9db01fdac05f1131fd2e930ef8b91128ce76b337f14R84), [index_describe_wait_for_ready](https://github.com/pinecone-io/cli/pull/32/files#diff-1531595eabd59249289ffd93383fc70230ac290ca5724c9fa6f6abd8790cec13R25), and [assert_index_doesnt_exist](https://github.com/pinecone-io/cli/pull/32/files#diff-1531595eabd59249289ffd93383fc70230ac290ca5724c9fa6f6abd8790cec13R90) would be more easily done in Go natively since the codebase and CLI are also written in Go. Additionally, we could use the Go SDK directly in test utilities alongside the CLI integration tests.

If we write the suite in bash, we also lose access to Go-specific testing frameworks and the `go test` command can no longer be relied on to run all related tests. When we were initially discussing the testing harness work, I should have been more specific in that it should be written in Go as well. I'm sorry for not clarifying.

At a high level, I think we need to lean into using Go to test the CLI binary directly. This should be doable through writing test code in Go that looks something like this:

```
func TestIndexList(t *testing.T) {
    bin := "./dist/pc_darwin_arm64/pc"
    cmd := exec.Command(bin, "index", "list")

    // Set env vars
    cmd.Env = append(os.Environ(), "PINECONE_API_KEY=testkey")

    var stdout, stderr bytes.Buffer
    cmd.Stdout = &stdout
    cmd.Stderr = &stderr

    err := cmd.Run()
    if err != nil {
        t.Fatalf("command failed: %v\nstderr: %s", err, stderr.String())
    }

    // assert on stdout and stderr
}
```

In the Go SDK, we use a dependency called [stretchr/testify](https://pkg.go.dev/github.com/stretchr/testify) for utilities around assertions and organizing test suites. We use suites heavily for our integration tests in the SDK codebase, and you can see some of that here: https://github.com/pinecone-io/go-pinecone/blob/main/pinecone/test_suite.go. I feel like what we should be shooting for in this repo is a top level folder called "integration" that holds all of the suite logic and individual test files for the built CLI binary. You could organize the files by command, etc:

```
├─ e2e/
│  ├─ e2e_test.go
│  ├─ project_test.go    # (pc project ...)
│  ├─ index_test.go       # (pc index …)
│  └─ api_key_test.go   # (pc api-key ...)  
```

I feel strongly that this is how we should be approaching testing for now. To be honest, I'm not much of a bash expert, although I have used it here and there. It was really hard for me to get my head around this testing framework to review, and the thought of needing to maintain and expand this going forward alongside developing the CLI seems like a big risk.

File: pcdev [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+                # Try goreleaser v2 naming first (pc_Linux_arm or pc_Linux_arm_v7)
+                if [ -f "$DIST_DIR/pc_Linux_arm_v7/pc" ]; then
+                    BINARY_PATH="$DIST_DIR/pc_Linux_arm_v7/pc"
+                elif [ -f "$DIST_DIR/pc_Linux_arm/pc" ]; then
+                    BINARY_PATH="$DIST_DIR/pc_Linux_arm/pc"
+                # Fallback to older naming conventions
+                elif [ -f "$DIST_DIR/pc_linux_arm_v7/pc" ]; then
+                    BINARY_PATH="$DIST_DIR/pc_linux_arm_v7/pc"
+                elif [ -f "$DIST_DIR/pc_linux_arm/pc" ]; then
+                    BINARY_PATH="$DIST_DIR/pc_linux_arm/pc"
```
Comment (austin-denoble) [inline pcdev:201]: Where did you get all these hard-coded paths for the different OS and arch variations? What version of goreleaser were you using?

I'm asking because these don't seem to really match up with what I see when I run things locally. I'm on v2.0.0, I think:

```
goreleaser --version
  ____       ____      _
 / ___| ___ |  _ \ ___| | ___  __ _ ___  ___ _ __
| |  _ / _ \| |_) / _ \ |/ _ \/ _` / __|/ _ \ '__|
| |_| | (_) |  _ <  __/ |  __/ (_| \__ \  __/ |
 \____|\___/|_| \_\___|_|\___|\__,_|___/\___|_|
goreleaser: Deliver Go Binaries as fast and easily as possible
https://goreleaser.com

GitVersion:    2.0.0
GitCommit:     f35dcda343ddebf9dae706c8a86aa5915501fa84
GitTreeState:  false
BuildDate:     2024-06-05T00:24:56Z
BuiltBy:       goreleaser
GoVersion:     go1.22.3
Compiler:      gc
ModuleSum:     h1:TxH7v+rYsQ+e4CQDMmdBFnVzxuTbBJRwsLXeCSkOREU=
Platform:      darwin/arm64
```

File: test/helpers/custom/indexes.bash [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+        echo "Error: JSON does not match template. Differences:" >&2
+        echo "$diff_output" >&2
+        return 1
+    fi
+}
+
+# Convert JSON template to CLI flags
+# Usage: cli_params=$(extract_cli_params_from_template "serverless_aws")
+# Returns: CLI flags string on success, error message on failure
+extract_cli_params_from_template() {
```
Comment (austin-denoble) [inline test/helpers/custom/indexes.bash:255]: This is quite a lot of bash to do something which is highly dependent on the shape of the index model and what we expect back from our APIs. Doing something like this in bash rather than natively in Go doesn't seem like the right thing to be doing in a test harness for our codebase right now.

File: test/install-deps.sh [hunk 10]
```diff
@@ -0,0 +1,266 @@
+#!/bin/bash
```
Comment (austin-denoble) [inline test/install-deps.sh:1]: Mentioned elsewhere, but I think we should be doing things like this through a justfile or something similar rather than in a complex bash file. I don't think it's very abnormal to just list out dependencies a developer may need to install as a part of the `CONTRIBUTING.md`, or even a just formula for bootstrapping.

If we keep our testing framework within the Go ecosystem, the management of dependencies can be handled by Go itself, which is a big advantage here too.

**Review by austin-denoble — COMMENTED** [review]
Another round of feedback primarily focused on the BATS framework and implementation. My apologies for not being really clear up front when we were discussing this, but we need to be building the testing suite in Go as well. This is a lot of very organized code, but it's really overwhelming, especially for someone who doesn't spend a lot of time in bash.

I'm also concerned because bash is notoriously finicky in various environments, shells, etc. Is BATS supported on Windows? I feel like the language itself is too fragile for the granularity and complexity we need here. The velocity factor in general development within the codebase is also important - we do not have many devs working in this repo, and the added complexity of a harness like this is going to be detrimental to velocity right now.
